### PR TITLE
Implement derive macro for EventSourced

### PR DESCRIPTION
**Pull request by ffuugoo**
Example:

```rust
#[derive(Event, EventSourced)]
#[event_sourced(aggregate = "User")]
enum UserEvents {
    UserCreated(UserCreated),
    UserRemoved(UserRemoved),
}
```

### CODE REVIEW

**Comment by ffuugoo** [comment]
### FCM

```
Implement derive macro for EventSourced

- reorganize crate implementations to maximize code reuse

Additionally:
- swap two paragraphs in derive-AggregateEvent documentation
```

**Comment by ffuugoo** [comment]
@tyranron ping

File: cqrs-codegen/impl/src/util.rs [hunk 1]
```diff
@@ -306,6 +333,61 @@ where
     )
 }
 
+pub(crate) fn get_type_if_type_parameter_used_in_type<'a>(
```
Comment (tyranron) [inline cqrs-codegen/impl/src/util.rs]: Docs?

File: cqrs-codegen/impl/src/util.rs [hunk 2]
```diff
@@ -306,6 +333,61 @@ where
...
+        match ty {
+            syn::Type::Reference(syn::TypeReference { elem: ty, .. }) => Some(ty),
+            ty => Some(ty),
+        }
+    } else {
+        None
+    }
+}
+
+fn is_type_parameter_used_in_type<'a>(
```
Comment (tyranron) [inline cqrs-codegen/impl/src/util.rs]: ...docs?

File: cqrs-codegen/impl/src/event_sourced.rs [hunk 3]
```diff
@@ -0,0 +1,144 @@
+//! Codegen for [`cqrs::Command`].
```
Comment (tyranron) [inline cqrs-codegen/impl/src/event_sourced.rs]: oops!

File: cqrs-codegen/impl/src/event_sourced.rs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+            predicates.push(syn::parse2(quote!(#aggregate: ::cqrs::EventSourced<#ty>))?);
+            predicates.push(syn::parse2(quote!(#ty: ::cqrs::Event))?);
+        }
+    }
+
+    Ok(quote! {
+        #[automatically_derived]
+        impl#impl_generics ::cqrs::EventSourced<#type_name#ty_generics> for #aggregate #where_clause {
+            #[inline(always)]
+            fn apply(&mut self, ev: &#type_name#ty_generics) {
```
Comment (tyranron) [inline cqrs-codegen/impl/src/event_sourced.rs]: @ffuugoo matching over `enum` with a lot of variants is not that trivial to be strictly inlined all the time. Here is the place where compiler should decide this, like we do in `derive(Event)` for `enum`s.

File: cqrs-codegen/impl/src/event_sourced.rs [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                match *ev {
+                    #body
+                }
+            }
+        }
+    })
+}
+
+/// Parses aggregate of [`cqrs::Command`] from `#[command(...)]` attribute.
+fn parse_event_sourced_aggregate(meta: &util::Meta) -> Result<String> {
```
Comment (tyranron) [inline cqrs-codegen/impl/src/event_sourced.rs]: Docs again... oops!

**Comment by ffuugoo** [comment]
My bad. Should be all fixed now.

**Pull request merged**